File: Iron, Magnet and Wall/FEMA2.py
# This solution exceeds the time limit. So try this in cpp! 
# problem link -> https://www.codechef.com/NOV20B/problems/FEMA2

for i in range(int(input())):
    n,k = map(int,input().split())
    s = input()
    mg = 0
    d = 0
    l = []
    q = []
    for i in range(n):
        if s[i] == "M" :
            for j in range(n):
                if s[j] == "I":
                    if "X" not in s[min(i,j):max(i,j)]:
                        if i not in l and j not in q:
                            p = k+1-abs(i-j)-s[min(i,j):max(i,j)].count(":")
                            if p>0:
                                d = d+1
                                l.append(i)
                                q.append(j)
                                break
                            elif j>i:
                                break

    print(d)

File: Iron, Magnet and Wall/FEMA2.cpp
#include <iostream>
#include <queue>

using namespace std;

int main() {
	int t;
	cin>>t;
	
	queue<int>mag;
	queue<int>iron;
	    
	while(t--){
	    int n,k=0,i=0,j=0,m=0;
	    cin>>n>>k;
	    string s;
	    cin>>s;
	    int p=k+1;
	    int l=0,r=0,count=0;
	
	    for(int i=0; i<n; i++){
	        if(s[i]=='M')
	        mag.push(i);
	        if(s[i]=='I')
	        iron.push(i);
	        
	        if(s[i]=='X' || i==n-1){
	            int l,r,sheet=0,q;
	            while(!mag.empty() && !iron.empty()){
	                sheet=0;
	                l=min(mag.front(), iron.front());
	                r=max(mag.front(), iron.front());
	                
	                for(q=l; q<=r; q++){
	                    if(s[q]==':')
	                    sheet++;
	                }
	                if((p-abs(l-r) -sheet)>0){
	                    count++;
	                    mag.pop();
	                    iron.pop();
	                }
	                
	                else if(mag.front() < iron.front()){
	                    mag.pop();
	                }
	                
	                else{
	                 iron.pop();   
	                }
	            }
	            
	            while(!mag.empty())
	            mag.pop();
	            
	            while(!iron.empty())
	            iron.pop();
	        }
	    }
	        cout <<count<< std::endl;
	}
	return 0;
}
